File: codes/daxpy.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#define N 20480
 
int main(void) {
 
  double *x, *y;
  size_t size = N*sizeof(double);
 
  x = (double *)malloc(size);
  y = (double *)malloc(size);
 
  // initialize x and y
  srand(time(NULL));
  double a = (double)random() / RAND_MAX;
  int i;
  for (i=0; i<N; i++)
    x[i] = (double)random() / RAND_MAX;
  for (i=0; i<N; i++)
    y[i] = (double)random() / RAND_MAX;
 
  // compute axpy on the host CPU cores
  double *yomp;
  yomp = (double *)malloc(size);
    #pragma omp parallel for
  for (i=0; i<N; i++) {
    yomp[i] = a * x[i] + y[i];
  }
 
  // compute axpy on the accelerator
  double *yacc;
  yacc = (double *)malloc(size);
#pragma acc kernels copyin(x[0:N],y[0:N]), copyout(yacc[0:N])
  for (i=0; i<N; i++) {
    yacc[i] = a*x[i] + y[i];
  }
 
  // verify the results
  double m = -1.;
  double tmp;
#pragma omp parallel for private(tmp) reduction(max:m)
  for (i=0; i<N; i++) {
    tmp = fabs( (yacc[i]-yomp[i])/yomp[i] );
    if ( tmp > m ) m = tmp;
  }
 
  // release memory
  free(x);
  free(y);
  free(yomp);
  free(yacc);
 
  if ( m < 1E-12 ) {
    printf("Success!\n");
    return 0;
  }
  else {
    printf("Failure!\n");
    return 1;
  }
}

File: codes/daxpy.f90
program main
    implicit none
 
    ! Size of vectors
    integer,parameter :: n = 20480
 
    real(8) :: a, m, tmp
    integer :: i
 
    real(8),dimension(:),allocatable :: x
    real(8),dimension(:),allocatable :: y
    real(8),dimension(:),allocatable :: yomp
    real(8),dimension(:),allocatable :: yacc
 
    call random_number(a)
 
    ! Allocate memory for each vector on host
    allocate(x(n))
    allocate(y(n))
    allocate(yomp(n))
    allocate(yacc(n))
 
    ! Initialize x and y
    call random_number(x)
    call random_number(y)
 
    ! Compute axpy on the host CPU cores
    !$omp parallel do
    do i=1,n
        yomp(i) = a*x(i) + y(i)
    enddo
    !$omp end parallel do
 
    ! Compute axpy on the accelerator
    !$acc kernels copyin(x(1:n),y(1:n)), copyout(yacc(1:n))
    do i=1,n
        yacc(i) = a*x(i) + y(i)
    enddo
    !$acc end kernels
 
    ! Verify the results
    m = -1.
    !$omp parallel do private(tmp) reduction(max:m)
    do i=1,n
        tmp = abs( (yacc(i)-yomp(i))/yomp(i) )
        if ( tmp > m ) m = tmp
    enddo
 
    if ( m < 1E-12) then
        print *, "Success!"
    else
       print *, "Failure!"
    endif
 
    ! Release memory
    deallocate(x)
    deallocate(y)
    deallocate(yomp)
    deallocate(yacc)
 
end program main
